src/nodes/mod.rs

```diff
@@ -88,57 +88,135 @@ pub struct AttrDelta {
+fn try_path<O: Fn(&PathBuf) -> nix::Result<()>>(path: Option<&PathBuf>, op: O) -> nix::Result<()> {
+    match path {
+        Some(path) => op(path),
+        None => Ok(()),
+    }
+}
+
+/// Helper function for `setattr` to apply only the mode changes.
+fn setattr_mode(attr: &mut fuse::FileAttr, path: Option<&PathBuf>, mode: Option<sys::stat::Mode>)
+    -> Result<(), nix::Error> {
+    if mode.is_none() {
+        return Ok(())
+    }
+    let mode = mode.unwrap();
 
-    if let Some(mode) = delta.mode {
-        result = result.and(
-            sys::stat::fchmodat(None, path, mode, sys::stat::FchmodatFlags::NoFollowSymlink));
+    let result = try_path(path, |p|
+        sys::stat::fchmodat(None, p, mode, sys::stat::FchmodatFlags::NoFollowSymlink));
+    if result.is_ok() {
+        attr.perm = mode.bits();
     }
+    result
+}
 
-    if delta.uid.is_some() || delta.gid.is_some() {
-        result = result.and(
-            unistd::fchownat(None, path, delta.uid, delta.gid,
-                unistd::FchownatFlags::NoFollowSymlink));
+/// Helper function for `setattr` to apply only the UID and GID changes.
+fn setattr_owners(attr: &mut fuse::FileAttr, path: Option<&PathBuf>, uid: Option<unistd::Uid>,
+    gid: Option<unistd::Gid>) -> Result<(), nix::Error> {
+    if uid.is_none() && gid.is_none() {
+        return Ok(())
     }
 
-    if delta.atime.is_some() || delta.mtime.is_some() {
-        let atime = delta.atime.unwrap_or_else(|| conv::timespec_to_timeval(attr.atime));
-        let mtime = delta.mtime.unwrap_or_else(|| conv::timespec_to_timeval(attr.mtime));
-        if attr.kind == fuse::FileType::Symlink {
-            // TODO(jmmv): Should use futimensat to avoid following symlinks but this function is
-            // not available on macOS El Capitan, which we currently require for CI builds.
-            warn!("Asked to update atime/mtime for symlink {} but following symlink instead",
-                path.display());
-        }
-        result = result.and(sys::stat::utimes(path, &atime, &mtime));
+    let result = try_path(path, |p|
+        unistd::fchownat(None, p, uid, gid, unistd::FchownatFlags::NoFollowSymlink));
+    if result.is_ok() {
+        attr.uid = uid.map_or_else(|| attr.uid, u32::from);
+        attr.gid = uid.map_or_else(|| attr.gid, u32::from);
     }
+    result
+}
+
+/// Helper function for `setattr` to apply only the atime and mtime changes.
+fn setattr_times(attr: &mut fuse::FileAttr, path: Option<&PathBuf>,
+    atime: Option<sys::time::TimeVal>, mtime: Option<sys::time::TimeVal>)
+    -> Result<(), nix::Error> {
+    if atime.is_none() && mtime.is_none() {
+        return Ok(());
+    }
+
+    let atime = atime.unwrap_or_else(|| conv::timespec_to_timeval(attr.atime));
+    let mtime = mtime.unwrap_or_else(|| conv::timespec_to_timeval(attr.mtime));
+    if attr.kind == fuse::FileType::Symlink {
+        // TODO(jmmv): Should use futimensat to avoid following symlinks but this function is
```

Comment:
Not a fan of this. We should at some point conditionalize and this path should fail if you try to do this to a symlink.